pkg/object/mutation/types.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2020 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package mutation
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cli-utils/pkg/object"
+)
+
+// ApplyTimeMutation is a list of substitutions to perform in the target
+// resource before applying, after waiting for the source resources to be
+// reconciled.
+// This most notibly allows status fields to be substituted into spec fields.
+type ApplyTimeMutation []FieldSubstitution
+
+// Equal returns true if the substitutions are eqivelent, ignoring order.
+// Fulfills Equal interface from github.com/google/go-cmp
+func (a ApplyTimeMutation) Equal(b ApplyTimeMutation) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	mapA := make(map[FieldSubstitution]struct{}, len(a))
+	for _, sub := range a {
+		mapA[sub] = struct{}{}
+	}
+	mapB := make(map[FieldSubstitution]struct{}, len(b))
+	for _, sub := range b {
+		mapB[sub] = struct{}{}
+	}
+	if len(mapA) != len(mapB) {
+		return false
+	}
+	for b := range mapB {
+		if _, exists := mapA[b]; !exists {
+			return false
+		}
+	}
+	return true
+}
+
+// FieldSubstitution specifies a substitution that will be performed at
+// apply-time. The source resource field will be read and substituted into the
+// target resource field, replacing the token.
+type FieldSubstitution struct {
+	// SourceRef is a reference to the resource that contains the source field.
+	SourceRef ResourceReference `json:"sourceRef"`
+
+	// SourcePath is a JSONPath reference to a field in the source resource.
+	// Example: "$.status.number"
+	SourcePath string `json:"sourcePath"`
+
+	// TargetPath is a JSONPath reference to a field in the target resource.
+	// Example: "$.spec.member"
+	TargetPath string `json:"targetPath"`
+
+	// Token is the substring to replace in the value of the target field.
+	// If empty, the target field value will be set to the source field value.
+	// Example: "${project-number}"
+	// +optional
+	Token string `json:"token,omitempty"`
+}
+
+// ResourceReference is a reference to a KRM resource by name and kind.
+// One of APIVersion or Group is required.
+// Group is generally preferred, to avoid needing to update the version in lock
+// step with the referenced resource.
+// If neither is provided, the empty group is used.
+type ResourceReference struct {
```

Comment:
The significant difference is supporting apiVersion OR apiGroup.

I wanted user flexibility. Sometimes you have one, sometimes you have the other. This avoids the user needing to parse and extract the group from the version. This would be valuable for cases like kpt-set that has one and not the other, avoiding needing to make a custom function to parse it.

But it does have a side effect that specifying the version is more restrictive. So if you think it's important I could get on board with only supporting apiGroup.

FWIW, I put some discussion of the inputs and alternatives and prior art in http://go/apply-time-mutation-in-kpt under the "KRM Resource References" header.

Reply:
ObjMetadata is exclusively used internally. ResourceReference is an external facing API.

Should they really be made the same thing?

Should the API not optionally accept APIVersion?

I'm thinking about renaming "group" to "apiGroup" to avoid confusion and be consistent with the [Cluster]RoleBinding and PersistentVolumeClaim APIs (tho others use "group"). If I do that, should that affect ObjMetadata too?

Reply:
I agree that keeping the ObjMetadata as an internal type and define a new one here makes sense. I'm a bit skeptical about supporting apiVersion here, as we are otherwise trying to avoid using the version. If I specify a version here, would we then also require that we fetch the resource from the apiserver with that exact version?

I think we have used just `Group` pretty consistently rather than `apiGroup`.

Reply:
I still think we should stick with `ObjMetadata`, but I will not block on this. Both of these types should not be part of the API, and they should not be exported. The API is the annotation--not these two reference types.

Reply:
It's useful to have the structs exported for clients (like kpt functions) to use when building the annotation yaml.